libs/features/mas/web-components/src/merch-datasource.js

```diff
@@ -148,60 +200,93 @@ const cache = new FragmentCache();
  */
 export class MerchDataSource extends HTMLElement {
     /**
-     * @type {import('./aem.js').AEM}
+     * @type {import('@adobe/mas-web-components').AEM}
      */
     #aem;
     cache = cache;
 
+    /**
+     * @type {HtmlElement[]}
+     */
+    refs = [];
+
     /**
      * @type {string} fragment path
      */
     path;
 
+    /**
+     * Consonant styling for CTAs.
+     */
+    consonant = false;
+
+    /**
+     * Internal promise to track the readiness of the web-component to render.
+     */
+    #readyPromise;
+
     static get observedAttributes() {
-        return ['source', 'path'];
+        return ['source', 'path', 'consonant'];
     }
 
     attributeChangedCallback(name, oldValue, newValue) {
         this[name] = newValue;
     }
 
     connectedCallback() {
+        this.consonant = this.hasAttribute('consonant');
+        this.clearRefs();
         const bucket =
             this.getAttribute(ATTR_AEM_BUCKET) ??
-            document.querySelector('mas-studio')?.getAttribute(ATTR_AEM_BUCKET);
+            'publish-p22655-e59341';
         this.#aem = new AEM(bucket);
-        this.fetchData();
+        this.refresh(true);
+    }
+
+    clearRefs() {
+        this.refs.forEach((ref) => {
+            ref.remove();
+        });
     }
 
-    refresh() {
-        this.cache.remove(this.path);
-        this.fetchData();
+    async refresh(flushCache = true) {
+        if (!this.path) return;
+
+        if (this.#readyPromise) {
+            const ready = await Promise.race([
```

Comment:
what does that do?

Reply:
i understand race & resolve but i don't understand the logic here

Reply:
@npeltier later in the file we have  `get updateComplete()` which is a promise that resolves when datasource job is done.
`readyPromise`  once pending, should not be re-initialized, otherwise some caller might end up awaiting for ever.
here, if we are fetching the data already, we skip refreshing.